Orange/misc/server_embedder.py

```diff
@@ -197,63 +250,75 @@ async def _encode_data_instance(
-        if data_bytes is None:
-            self.num_parallel_requests -= 1
-            return None
-
-        # if data in cache return it
-        cache_key = self._cache.md5_hash(data_bytes)
-        emb = self._cache.get_cached_result_or_none(cache_key)
-
-        if emb is None:
-            # in case that embedding not sucessfull resend it to the server
-            # maximally for MAX_REPEATS time
-            for i in range(1, self.MAX_REPEATS + 1):
-                self.__check_cancelled()
+        while not queue.empty():
+            # remove in 3.33
+            self.__check_cancelled()
+
+            # get item from the queue
+            i, data_instance, num_repeats = await queue.get()
+            num_repeats += 1
+
+            # load bytes
+            data_bytes = await self._encode_data_instance(data_instance)
+            if data_bytes is None:
+                continue
+
+            # retrieve embedded item from the local cache
+            cache_key = self._cache.md5_hash(data_bytes)
+            log.debug("Embedding %s", cache_key)
+            emb = self._cache.get_cached_result_or_none(cache_key)
+
+            if emb is None:
+                # send the item to the server for embedding if not in the local cache
+                log.debug("Sending to the server: %s", cache_key)
                 url = (
-                    f"/{self.embedder_type}/{self._model}?"
-                    f"machine={self.machine_id}"
-                    f"&session={self.session_id}&retry={i}"
+                    f"/{self.embedder_type}/{self._model}?machine={self.machine_id}"
+                    f"&session={self.session_id}&retry={num_repeats}"
                 )
                 emb = await self._send_request(client, data_bytes, url)
                 if emb is not None:
                     self._cache.add(cache_key, emb)
-                    break  # repeat only when embedding None
-        if proc_callback:
-            proc_callback(emb is not None)
 
-        self.num_parallel_requests -= 1
-        return emb
+            if emb is not None:
+                # store result if embedding is successful
+                log.debug("Successfully embedded:  %s", cache_key)
+                results[i] = emb
+                if proc_callback:
+                    proc_callback()
+            elif num_repeats < self.MAX_REPEATS:
+                log.debug("Embedding unsuccessful - reading to queue:  %s", cache_key)
+                # if embedding not successful put the item to queue to be handled at
+                # the end - the item is put to the end since it is possible that  server
+                # still process the request and the result will be in the cache later
+                # repeating the request immediately may result in another fail when
+                # processing takes longer
+                queue.put_nowait(TaskItem(i, data_instance, no_repeats=num_repeats))
+            queue.task_done()
```

Comment:
Shouldn't task_done be called only when the data_instance was correctly processed, so in the `if emb is not None` part? Or do I miss something?

Reply:
Oh, no, my mistake: I missed that an element is put into the queue beforehand if the computation was not successful, so ignore the previous comment.

Reply:
As you noticed it was not successful processed but the new item is in the queue and the current one is done.